### DISCUSSION TITLE
Attribute handling

### DISCUSSION THREAD

**Discussion by ramcdona** [BrunoLevy/geogram#129] [General]
What happens to attributes when various operations are done on meshes?

In particular, I have two cases I care about -- and I am mostly concerned with Boolean (like) operations.

ID's of faces -- every triangle may have a unique (integer) ID number, or a group may share a groupID.  When a triangle is split (during the intersection process), I would like the sub-triangles to inherit the value of the parent ID and groupID.

U,V parametric coordinates of vertices -- vertex viewed from a face -- every vertex will have (floating point) U,V coordinates.  When a triangle is split (during the intersection process), I would like the new vertices to be assigned interpolated U,V coordinate values.  Vertices along an intersection come from both MeshA and MeshB, so while there would be a single XYZ, I would want those verts to have a U_A,V_A and U_B,V_B.

I can imagine other cases where different behavior may be desired -- depending on the meaning of the quantity and the operation being performed.  Since no single answer fits for all purposes, a flexible system probably allows choice.  Perhaps Geogram provides a callback function to handle this situation...

**Comment by BrunoLevy** [comment]
Yes, geogram has a generic attribute handling mechanism, with generic attribute interpolation that will work for interpolating UV coordinates and colors. All the wires at not plugged yet in the mesh intersection code, but it is doable:
- for attributes attached to facets, the wires are plugged (each time a facet is subdivided, its attributes are copied to the new facets)
- for attributes attached to vertices, the wires remain to be plugged, but it will not be super difficult to do: each time a facet is retriangulated, the attribute interpolator can be called on the newly generated vertices. There is just a rather annoying subtlety regarding UV coordinates that can be attached to vertices or to facet corners (facet corners are more difficult to handle properly...)

**Reply by ramcdona** [reply]
Thanks for your response.  I finally have the opportunity to return to this.  I am trying to incorporate attributes into my program.

I believe I'm doing things right, but the attributes are not there when I go to retrieve them.  Do you have an example program that does boolean operations on meshes with attributes?

**Reply by ramcdona** [reply]
I've upgraded from 1.8.6 to the tip of GitHub.  It appears that attributes are not propagated through Boolean operations.  Please let me know if there is something I must to do activate them.

The attributes themselves come through -- but their values are universally zero.

**Reply by BrunoLevy** [reply]
Boolean operations are still under development, and attribute propagation / interpolation is not implemented yet. If you want to beta-test it, I can add it (will be there in 1 or 2 weeks probably). What do you need in priority ? vertex attributes or facet corner attributes ? (if it is vertex attributes, then it is easier, facet corner attributes are more tricky). 

If you want to add it on your own, you'll need to dive into the code, it is mostly [here](https://github.com/BrunoLevy/geogram/blob/8da098632ebbd74f2c4ee14f0f50ef4061ad2d1f/src/lib/geogram/mesh/mesh_surface_intersection_internal.cpp#L360)
(newly created points need their attributes to be interpolated). The interpolation weights can be deduced from the combinatorial information of the new point, as done for the geometry [here](https://github.com/BrunoLevy/geogram/blob/8da098632ebbd74f2c4ee14f0f50ef4061ad2d1f/src/lib/geogram/mesh/mesh_surface_intersection_internal.cpp#L152) and that's mostly all that needs to be done for vertex attributes. Now if you have corner attributes, it is a bit trickier: for each newly created vertex, you need to propagate the attributes to all facet corners that touch the vertex, and you need also to keep everything in sync when retriangulating the co-planar facets [here](https://github.com/BrunoLevy/geogram/blob/8da098632ebbd74f2c4ee14f0f50ef4061ad2d1f/src/lib/geogram/mesh/mesh_surface_intersection_internal.cpp#L989).

Since it's a bit complicated, I have kept it for when I'll be completely happy with the intersection code (not completely the case, I'm now working on handling some "corner cases", with some non-manifold "fins" floating around). When I'm sufficiently done, I'll add attribute management for vertex attributes, and then only for facet corner attributes.

**Reply by ramcdona** [reply]
For my application, I need both facet face attributes (perhaps the simplest case of all, no interpolation -- just propagation) and facet corner attributes.

However, my facet corner attributes are perhaps a little closer to vertex attributes than the general case.  My facet corner attributes are U,V surface coordinates.  When I start with a single surface, they are equivalent to vertex attributes.  However, once intersected with another surface, the points along the intersection will have U,V values from surface 1 and also U,V values from surface 2.

There are (I believe) two other features I need to figure out that might not be complete in Geogram.

1) Discussed [here](https://github.com/BrunoLevy/geogram/discussions/128).  Be able to work with non-manifold 'thin' surfaces through Boolean-like operations.  These already seem to work pretty well for the simple case, but it is not documented and I am not sure if the results I'm seeing are expected.  

2) Another "new" Boolean operation -- to Keep all of A.  I.e. intersect A with B, (generate intersection curve, re-triangulate both sides, etc.) and then keep all of surface A, discard all of B.  This will leave A with an imprint of the intersection with B.

I believe 1 & 2 can be completed with appropriate generalizations of the `classify()` function, but upon first reading, it is well beyond my understanding.


As for priority (between these items), it is difficult to say.  I really need all these features before I can use Geogram to replace our existing code.  Hopefully that could be on a timeframe of less than six months from now (sooner is of course always better).   Knowledge that these issues are being worked on of course goes a long way and is much appreciated.

I will have a look at the code segments you reference above.  However, considering how lost I got in `classify()`, I am not comfortable having these changes 'assigned' to me at this time.  I will let you know if I think I can handle it.

Thanks again for all the work.

**Reply by ramcdona** [reply]
Digging in a little, I see that vertex attributes could be interpolated in a few ways.  As you point out, most of the code is [here](https://github.com/BrunoLevy/geogram/blob/8da098632ebbd74f2c4ee14f0f50ef4061ad2d1f/src/lib/geogram/mesh/mesh_surface_intersection_internal.cpp#L152).

Modifying mix(t,p1,p2) to be attribute aware would almost suffice.  Alternately, returning (as a reference) t from plane_line_intersection(..., exact::rational & t) would expose the interpolation parameter that is needed.

What is missing from the above code snippets is a canonical means of iterating over all attributes and adding them for newly created vertices.

As for faces, it appears that [this line](https://github.com/BrunoLevy/geogram/blob/8da098632ebbd74f2c4ee14f0f50ef4061ad2d1f/src/lib/geogram/mesh/mesh_surface_intersection_internal.cpp#L395) should do what is required.  However, in my tests, it doesn't seem to work.  Perhaps I am doing something wrong.

**Reply by BrunoLevy** [reply]
Hi !

- there used to be a generic attribute interpolation mechanism, it was more complicated than necessary, so I removed it, but I will resurect (a simpler form of) it !

- the easiest way to plug it will be in `MeshInTriangle::commit()`, this is where new vertices are created in the mesh. For each new vertex, we know in which triangle it was created, and we need to interpolate from the three vertices of the initial triangle. We can easily recompute the interpolation weights (I think that recomputing them will cost less than propagating them all the way from the intersection functions), they are just ratios of triangle areas. Moreover, they can be computed with standard doubles (no need for arbitrary precision arithmetics here).

**Reply by BrunoLevy** [reply]
I have been thinking about it, now I have a plan (see issue #132), will work on it asap.

**Reply by BrunoLevy** [reply]
Generic attribute interpolator flies !
Now working on plugging all the wires in the mesh intersection algorithm... (facet corners attributes are harder, but not that hard in fact).

**Reply by ramcdona** [reply]
Thanks for jumping on this.  Much appreciated.

Let me know when you'd like me to test it out and I'll try to get right on it.

**Reply by BrunoLevy** [reply]
You can try it (get latest version from the repository)
Before computing the intersection, call `MeshSurfaceIntersection::set_interpolate_attributes(true)`. This will interpolate all facet corner attributes. 
Note: vertex attribute are not handled, because an intersection vertex is on several facets so would have possibily different values for the same attribute (but if you have vertex attributes in your input mesh, you can copy them to facet corner attributes).
It works for attributes of type double, float, int32_t, uint32_t.

**Reply by BrunoLevy** [reply]
![image](https://github.com/BrunoLevy/geogram/assets/19478253/489eb0fc-9601-4041-9cfb-5fecc2224457)
screenshot: intersection of a parameterized mesh (u,v coordinates attached to facet corners)

**Reply by ramcdona** [reply]
I will try it out soon and let you know how it goes -- I'm sure it will work great.

That UV visualization is great -- do you texture map a grid to do it, or something else?

**Reply by BrunoLevy** [reply]
It is simply a texture mapped in repeat mode. It is in the Graphite software (based on geogram, also on github)

**Reply by ramcdona** [reply]
Thanks, that is what I figured.  I started trying to think of how I would accomplish the same -- and then I thought of using a texture and just letting the graphics card worry about it.  Nice trick.

**Reply by ramcdona** [reply]
Thanks again for your progress on this.

Can you post the sample code you used to create the above intersection (don't worry about visualization code).

For example, I don't see how to `MeshSurfaceIntersection::set_interpolate_attributes(true);` -- `set_interpolate_attributes` is not a static member function, yet (so far) I do not instantiate a `MeshSurfaceIntersection` class.

I am currently working with `GEO::mesh_union()` as my entry point to CSG type operations including surface intersection.  Is there a different way I should be working?

**Reply by ramcdona** [reply]
OK, I've gone back into the code for mesh_boolean_operation -- I see a path forward now.  Thanks.

**Reply by ramcdona** [reply]
I spent about a day wrestling with it and couldn't get anything working right.  The built-in orig_face_id mostly seems to work, but I haven't been able to use it for what I want.

I think I'm going to need to make co-planar merging less aggressive.  I will disable merging of faces with different ID's.

Although MeshInTriangle::commit() has `target_mesh().facets.attributes().copy_item(new_t, f1_);` I seem unable to get other facet attributes to work.

Similarly with facet corner attributes -- I can set them up, but they come out the other end as 0.0.

I am going to be at a conference for the next week and will not have any time to keep playing with it.  However, I believe I am doing something stupid or missing something simple.  I suspect if you could share an example program, it would be enough to get me sorted.

Thanks again.

**Reply by BrunoLevy** [reply]
OK, I'll write a small demo program with attributes so that it will make things easier.

About merging coplanar facets, for the moment, just ignore it (do not call this function). I guess merging facets with same original facet id will not make something very different from not merging them since they are re-triangulated. It may gain something if you got a mesh with non-triangular facets (but then we need to keep original facet idea in another attribute since 'original_facet_id' is used by the algorithm for internal bookkeeping)

**Reply by ramcdona** [reply]
Have you had an opportunity to put together such a demo program?  If not, do you have any other tips or pointers that could put me in the right direction here?

**Comment by ramcdona** [comment]
It seems that attributes 'don't work' through `mesh_boolean_operation()` because whatever attributes exist in the operands, they are not copied into the `results` mesh by `copy_operand()` [here](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram/mesh/mesh_surface_intersection.cpp#L2439) before processing gets going.

So, by the time `intersect_epilogue()` is called and could interpolate attributes, there are no attributes to work with.

I tried adding

```
result.facets.attributes().copy( operand.facets.attributes() );
result.facet_corners.attributes().copy( operand.facet_corners.attributes() );
```

At the end of `copy_operand()`, but that does not really seem to do what is needed here.  All the other `copy_item()` related calls in `AttributesManager` seem to only deal with manipulations within a single `AttributesManager` -- whereas we need to copy the attributes from one mesh (and therefore one AM) to another.

It seems like I could handle this transfer very manually via `list_attribute_names()` and then for-looping over each name, with a bunch of direct work with indexes into `faces` and `face_corners` manually kept track of.  Part of the complication is that we need to copy multiple operands into one mesh -- so the attributes need to be properly appended into the data structure (not copied over).
